File: src/convert.js
import yaml from 'js-yaml';
import { parseCsv, stringifyCsv } from './csv.js';

/**
 * Detect format from filename extension.
 * @param {string} [filename]
 * @returns {'json' | 'yaml' | 'csv' | null}
 */
export function detectFormat(filename) {
  if (!filename) return null;
  const lower = filename.toLowerCase();
  if (lower.endsWith('.json')) return 'json';
  if (lower.endsWith('.yaml') || lower.endsWith('.yml')) return 'yaml';
  if (lower.endsWith('.csv')) return 'csv';
  return null;
}

/**
 * Parse input text as the given format.
 * @param {string} text
 * @param {'json' | 'yaml' | 'csv'} format
 * @returns {unknown}
 */
export function parse(text, format) {
  switch (format) {
    case 'json':
      return JSON.parse(text);
    case 'yaml':
      return yaml.load(text);
    case 'csv':
      return parseCsv(text);
    default:
      throw new Error(`Unsupported parse format: ${format}`);
  }
}

/**
 * Serialize data to the given format.
 * @param {unknown} data
 * @param {'json' | 'yaml' | 'csv'} format
 * @param {{ pretty?: boolean, minify?: boolean }} [opts]
 * @returns {string}
 */
export function serialize(data, format, opts = {}) {
  switch (format) {
    case 'json': {
      if (opts.minify) return JSON.stringify(data);
      const space = opts.pretty === false ? 0 : 2;
      return JSON.stringify(data, null, space) + (space ? '\n' : '');
    }
    case 'yaml':
      return yaml.dump(data, { lineWidth: -1, noRefs: true });
    case 'csv': {
      if (!Array.isArray(data)) {
        throw new Error('CSV output requires an array (of objects or rows)');
      }
      return stringifyCsv(data);
    }
    default:
      throw new Error(`Unsupported serialize format: ${format}`);
  }
}

/**
 * Convert between formats.
 * @param {string} text
 * @param {'json' | 'yaml' | 'csv'} from
 * @param {'json' | 'yaml' | 'csv'} to
 * @param {{ pretty?: boolean, minify?: boolean }} [opts]
 * @returns {string}
 */
export function convert(text, from, to, opts = {}) {
  const data = parse(text, from);
  return serialize(data, to, opts);
}

/**
 * Infer source format from content heuristics when extension is unknown.
 * @param {string} text
 * @returns {'json' | 'yaml' | 'csv'}
 */
export function inferFormat(text) {
  const trimmed = text.trimStart();
  if (!trimmed) return 'json';
  if (trimmed[0] === '{' || trimmed[0] === '[') {
    try {
      JSON.parse(text);
      return 'json';
    } catch {
      // fall through
    }
  }
  const firstLine = trimmed.split(/\r?\n/, 1)[0] ?? '';
  if (firstLine.includes(',') && !trimmed.startsWith('---') && !/:\s/.test(firstLine)) {
    return 'csv';
  }
  return 'yaml';
}

File: src/data_transform/convert.py
"""Format detection, parse, serialize, and convert helpers."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Literal

import yaml

from .csv_io import parse_csv, stringify_csv

FormatName = Literal["json", "yaml", "csv", "ndjson"]


def detect_format(filename: str | None) -> FormatName | None:
    """Detect format from filename extension."""
    if not filename:
        return None
    lower = str(filename).lower()
    # Support bare names and paths.
    name = Path(lower).name
    if name.endswith(".ndjson") or name.endswith(".jsonl"):
        return "ndjson"
    if name.endswith(".json"):
        return "json"
    if name.endswith(".yaml") or name.endswith(".yml"):
        return "yaml"
    if name.endswith(".csv"):
        return "csv"
    return None


def parse(text: str, format: FormatName) -> Any:
    """Parse input text as the given format."""
    if format == "json":
        return json.loads(text)
    if format == "yaml":
        return yaml.safe_load(text)
    if format == "csv":
        return parse_csv(text)
    if format == "ndjson":
        return _parse_ndjson(text)
    raise ValueError(f"Unsupported parse format: {format}")


def _parse_ndjson(text: str) -> list[Any]:
    """Parse newline-delimited JSON into a list of values."""
    values: list[Any] = []
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        values.append(json.loads(stripped))
    return values


def serialize(
    data: Any,
    format: FormatName,
    *,
    pretty: bool | None = None,
    minify: bool = False,
) -> str:
    """Serialize data to the given format."""
    if format == "json":
        if minify:
            return json.dumps(data, separators=(",", ":"), ensure_ascii=False)
        space = 0 if pretty is False else 2
        body = json.dumps(data, indent=space or None, ensure_ascii=False)
        if space:
            return body + "\n"
        return body
    if format == "yaml":
        return yaml.safe_dump(
            data,
            sort_keys=False,
            allow_unicode=True,
            default_flow_style=False,
            width=10**9,
        )
    if format == "csv":
        if not isinstance(data, list):
            raise ValueError("CSV output requires an array (of objects or rows)")
        return stringify_csv(data)
    if format == "ndjson":
        return _serialize_ndjson(data)
    raise ValueError(f"Unsupported serialize format: {format}")


def _serialize_ndjson(data: Any) -> str:
    """Emit compact NDJSON: one JSON value per line."""
    if isinstance(data, list):
        lines = [
            json.dumps(item, separators=(",", ":"), ensure_ascii=False) for item in data
        ]
        return ("\n".join(lines) + "\n") if lines else ""
    return json.dumps(data, separators=(",", ":"), ensure_ascii=False) + "\n"


def convert(
    text: str,
    from_format: FormatName,
    to_format: FormatName,
    *,
    pretty: bool | None = None,
    minify: bool = False,
) -> str:
    """Convert between formats."""
    data = parse(text, from_format)
    return serialize(data, to_format, pretty=pretty, minify=minify)


def infer_format(text: str) -> FormatName:
    """Infer source format from content heuristics when extension is unknown."""
    trimmed = text.lstrip()
    if not trimmed:
        return "json"
    # Prefer NDJSON when multiple non-empty lines each look like JSON values.
    if _looks_like_ndjson(trimmed):
        return "ndjson"
    if trimmed[0] in "{[":
        try:
            json.loads(text)
            return "json"
        except json.JSONDecodeError:
            pass
    first_line = trimmed.splitlines()[0] if trimmed else ""
    if (
        "," in first_line
        and not trimmed.startswith("---")
        and not re.search(r":\s", first_line)
    ):
        return "csv"
    return "yaml"


def _looks_like_ndjson(text: str) -> bool:
    """True when two or more non-empty lines each parse as a JSON object/array."""
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    if len(lines) < 2:
        return False
    for line in lines:
        if not line or line[0] not in "{[":
            return False
        try:
            val = json.loads(line)
        except json.JSONDecodeError:
            return False
        if not isinstance(val, (dict, list)):
            return False
    return True
